future-apps/payroll/contracts/Payroll.sol

```diff
@@ -594,21 +595,54 @@ contract Payroll is EtherTokenConstant, IForwarder, IsContract, AragonApp {
+     * @param _employeeId Employee's identifier.
+     * @param _payedAmount Amount payed to the employee to query the timestamp of.
+     * @return The date timestamp in seconds corresponding to the given paying amount based on the employee's last payroll date.
+     */
+    function _getLastPayroll(uint256 _employeeId, uint256 _payedAmount) internal view returns (uint64) {
+        Employee storage employee = employees[_employeeId];
+
+        uint256 timeDiff = _payedAmount.div(employee.denominationTokenSalary);
+
+        // This function is only called from _payday, where we make sure that _payedAmount is lower or equal to the
+        // total owed amount, that is obtained from _getOwedSalary, which does exactly the opposite calculation:
+        // multiplying the employee's salary by an uint64 number of seconds. Therefore, timeDiff will always fit in 64.
+        // Nevertheless, we are performing a sanity check at the end to ensure the computed last payroll timestamp
+        // is not greater than the current timestamp.
+
+        uint256 lastPayrollDate = employee.lastPayroll.add(timeDiff);
+        require(lastPayrollDate <= uint256(getTimestamp64()), ERROR_LAST_PAYROLL_DATE_TOO_BIG);
+        return uint64(lastPayrollDate);
+    }
+
+    /**
+     * @dev Gets amount of owed salary for a given employee until now.
+     * @param _employeeId Employee's identifier.
+     * @return Total amount of owed salary for the requested employee until now.
+     */
+    function _getOwedSalary(uint256 _employeeId) internal view returns (uint256) {
         Employee storage employee = employees[_employeeId];
 
+        // Get the min of current date and termination date
+        uint64 date = _isEmployeeActive(_employeeId) ? getTimestamp64(): employee.endDate;
+
         // Make sure we don't revert if we try to get the owed salary for an employee whose start
-        // date is in the future (necessary in case we need to change their salary before their
-        // start date)
-        if (_date <= employee.lastPayroll) {
+        // date is in the future (necessary in case we need to change their salary before their start date)
+        if (date <= employee.lastPayroll) {
             return 0;
         }
 
-        // Get time that has gone by (seconds)
-        // No need to use safe math as the underflow was covered by the previous check
-        uint64 time = _date - employee.lastPayroll;
+        // Get time diff in seconds, no need to use safe math as the underflow was covered by the previous check
+        uint64 timeDiff = date - employee.lastPayroll;
+        uint256 result = employee.denominationTokenSalary * uint256(timeDiff);
 
-        return employee.denominationTokenSalary.mul(time);
+        // Return max int if the result overflows
+        if (result / timeDiff != employee.denominationTokenSalary) {
+            return MAX_UINT256;
```

Comment:
I don't know if this would work that well in practice; if you use `_addAccruedValue()` afterwards (like we do in `setEmployeeSalary()`), you'd still revert there if there was any existing accrued value.

I think more consistent behaviour would be to just use safe math, or possibly even revert with better error strings for overflowing salary or accrued value.

Reply:
Discussed offline: we will keep the `MAX_UINT256` capping here to avoid employees from cashing out their salary if they reached a huge owed amount.